taichi/lang_util.h

```diff
@@ -19,10 +19,59 @@ struct Context;
 
 using FunctionType = std::function<void(Context &)>;
 
-enum class DataType : int {
+class Type {
+ public:
+  virtual std::string serialize() const = 0;
+  virtual ~Type() {
+  }
+};
+
+// A "Type" handle. This should be removed later.
+class DataType {
+ public:
+#define PER_TYPE(x) static DataType x;
+#include "taichi/inc/data_type.inc.h"
+#undef PER_TYPE
+  DataType() : ptr_(unknown.ptr_) {
+  }
+
+  DataType(const Type *ptr) : ptr_(ptr) {
+  }
+
+  bool operator==(const DataType &o) const {
+    return ptr_ == o.ptr_;
+  }
+
+  bool operator!=(const DataType &o) const {
+    return !(*this == o);
+  }
+
+  operator std::size_t() const;
+
+  std::string serialize() const {
```

Comment:
I didn't follow the history of this PR. But do I understand correctly that we initialially wanted to use `serialize()` for pickling, but then switched to `operator size_t()`..?

Sorry I don't have any experience on mypy. One thing I'd suggest considering is to make sure serializing/deserializing (or pickle) to go through the same interface... Right now we use `DataType` for picking, but `PrimitiveType` for unpickling. I guess in the end, the pickled bytes need to encode the information that allows the unpickle procedure to be able to dynamcially dispatch to the right "Type Node"?

E.g. if we have a pickle format of `"<type>/<enum>"`. For `PrimitiveType`, the serialized type could look like `"prm/2"`, where  `2` maps to `PrimitiveType`'s `i32`. Then during unpickling, by looking at `"prm"`, the type system could could dispatch to `PrimitiveType`, which could then take in `"2"` and returns us `*Primitive::i32`. So we probably need some sort of type registry to handle the `"<type>"` part..

Reply:
> I didn't follow the history of this PR. But do I understand correctly that we initialially wanted to use serialize() for pickling, but then switched to operator size_t()..?

I am not sure what `serialize()` is used for. Using `operator size_t()` is just my choice as it is perfect for handling the pickling problem, it originally should serve other purposes I think.

> Right now we use DataType for picking, but PrimitiveType for unpickling. 

This is not very precise as `DataType` is a handler for `PrimitiveType`, during pickling and unpickling of `DataType`, the procedure always works on the underlying `PrimitiveType` stored in the `DataType`.

Reply:
> , the procedure always works on the underlying PrimitiveType stored in the DataType.

That's the problem I was talking about. If `Type` is a virtual base class, it implies that we will have concrete implementations other than `PrimitiveType`. When that is the case, the unpickling procedure cannot just assume that it will work on `PrimitiveType`. Instead, it needs to figure out which impl to use by looking at the pickled data.

E.g. say if we have `PrimitiveType` and `FooType`. During unpickling, how do we know if it is `PrimitiveType`'s or `FooType`'s pointers that we need? Eventually, such information would need to come from the pickled data, right?